### PR TITLE
DummyParticles: Main: Mark particles as deleted on deletion.

### PR DESCRIPTION
**Pull request by SteffenSeckler**
# Description

- [x] removed RMM because why not...
- [x] don't actually delete particles, but instead, mark them as dummy/deleted.
- [x] [Jenkinsfile] remove threadsanitizer because it is fully covered by archer
- [x] dummy particles need to be removed on a rebuild of the container.
- [x] Fixes incorrect handling of masks in `LJFunctorAVX::SoAKernel()`. `upot` was used instead of the masked version (only for non-cell-wise-owned-state)
- Adapt functors to ignore dummy particles:
   - [x] LJFunctor
   - [x] LJFunctorAVX
   - [x] LJFunctorCuda
   - [x] LJFunctorCudaGlobals
   - [ ] FlopCounterFunctor <-- do we need that? I don't really care about it...
   - [x] SPHDensityFunctor
   - [x] SPHForceFunctor
- [x] Deleting particles using `AutoPas::deleteParticle` no longer invalidates the neighbor lists.

## Resolved Issues

- Needed for #434 (parallel_for, ...)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

- [x] Test if `updateContainer()` removes dummy particles.
- [x] Test if newly deleted particles are not still used with SoA <-- This is tested through the test `TraversalComparison`. While the particles are still inside of the SoA, they are explicitly skipped inside of the SoAFunctors.
- [x] Test with dummies to LJFunctorAVXTest

### CODE REVIEW

**Comment by lunaticcoding** [comment]
@FG-TUM @SteffenSeckler Wie schauts mit dem branch aus? :S I bin von dem geblockt :S isch des soweit, dass i n in meinen rein mergen kann? :)

**Comment by SteffenSeckler** [comment]
awkward language you are speaking 😋
There isn't missing much. All functors except for the LJFunctor need adaption for the new things, so it will currently only compile if you disable AVX. You can manually set the vectorization either to off or to SSE, then it should work.
So you can merge it to your branch, but the tests will most likely fail.

Or you can wait until @FG-TUM adapts the LJFunctorAVX...

Cuda stuff is of course also not adapted, but well...

**Comment by lunaticcoding** [comment]
@SteffenSeckler Sorry ;D My bad. I think I'll just merge it now and implement my stuff and then wait for @FG-TUM to wrap it up before diving into testing :) thanks :)

**Comment by SteffenSeckler** [comment]
@lunaticcoding you can merge this again into your branch. most of the problems regarding avx etc. should be fixed. this also includes a fix for both verletcluster containers with enabled newton3.

**Comment by SteffenSeckler** [comment]
@FG-TUM do we still need the FlopCounterFunctor? And if yes, should this be adapted or not?

**Comment by SteffenSeckler** [comment]
@lunaticcoding this PR now includes the necessary changes for cuda. Only SPH is missing at this point (which can be found in #494 )

File: src/autopas/containers/verletClusterLists/Cluster.h [hunk 1]
```diff
@@ -49,6 +49,31 @@ class Cluster {
    */
   const Particle &operator[](size_t index) const { return *(_firstParticle + index); }
 
+  /**
+   * Get Minimum and Maximum of the particles in z-direction.
+   * @return Pair of minimum and maximum in z-direction.
+   */
+  std::pair<double, double> getZMinMax() const {
```
Comment (FG-TUM) [inline src/autopas/containers/verletClusterLists/Cluster.h:56]:
```suggestion
  [[nodiscard]] std::pair<double, double> getZMinMax() const {
```

File: src/autopas/molecularDynamics/LJFunctor.h [hunk 2]
```diff
@@ -283,14 +269,16 @@ class LJFunctor
 
         const SoAFloatPrecision dr2 = drx2 + dry2 + drz2;
 
-        const SoAFloatPrecision mask = (dr2 > cutoffsquare) ? 0. : 1.;
+        // ownedStateI is already checked.
+        const bool mask =
+            dr2 <= cutoffsquare and /*ownedStateI != OwnershipState::dummy and */ ownedStateJ != OwnershipState::dummy;
```
Comment (FG-TUM) [inline src/autopas/molecularDynamics/LJFunctor.h:274]:
```suggestion
        // mask away if distance is too large or any particle is a dummy. Particle ownedStateI was already checked previously.
        const bool mask =
            dr2 <= cutoffsquare ownedStateJ != OwnershipState::dummy;
```

File: src/autopas/molecularDynamics/LJFunctor.h [hunk 3]
```diff
@@ -309,25 +297,16 @@ class LJFunctor
...
-            // The calculations are later divided by two!
-            double energyFactor = isOwnedI + ownedPtr[j];
-            upotSum += upot * energyFactor;
-            virialSumX += virialx * energyFactor;
-            virialSumY += virialy * energyFactor;
-            virialSumZ += virialz * energyFactor;
-          }
+          const SoAFloatPrecision upot = mask ? (epsilon24 * lj12m6 + shift6) : 0.;
+
+          // In this functor, all pairs are only traversed once (newton3-scheme!).
```
Comment (FG-TUM) [inline src/autopas/molecularDynamics/LJFunctor.h:302]:
```suggestion
          // In this function, all pairs are only traversed once (newton3-scheme!).
```

File: src/autopas/molecularDynamics/LJFunctor.h [hunk 4]
```diff
@@ -353,41 +329,26 @@ class LJFunctor
 
   // clang-format off
```
Comment (FG-TUM) [inline src/autopas/molecularDynamics/LJFunctor.h:330]: is this still needed here?

File: src/autopas/molecularDynamics/LJFunctor.h [hunk 5]
```diff
@@ -483,14 +434,16 @@ class LJFunctor
 
         const SoAFloatPrecision dr2 = drx2 + dry2 + drz2;
 
-        const SoAFloatPrecision mask = (dr2 > cutoffsquare) ? 0. : 1.;
+        // ownedStateI is already checked.
+        const bool mask =
+            dr2 <= cutoffsquare and /*ownedStateI != OwnershipState::dummy and */ ownedStateJ != OwnershipState::dummy;
```
Comment (FG-TUM) [inline src/autopas/molecularDynamics/LJFunctor.h:439]: same as above

File: src/autopas/molecularDynamics/LJFunctor.h [hunk 6]
```diff
@@ -937,13 +855,10 @@ class LJFunctor
...
     const size_t *const __restrict__ neighborListPtr = neighborList.data();
 
-    // checks whether particle 1 is in the domain box, unused if _duplicatedCalculations is false!
-    SoAFloatPrecision inbox1Mul = 0.;
-    if (duplicatedCalculations) {  // only for duplicated calculations we need this value
-      inbox1Mul = ownedPtr[indexFirst];
-      if (newton3) {
-        inbox1Mul *= 0.5;
-      }
+    // checks whether particle 1 is in the domain box, unused if calculateGlobals is false!
```
Comment (FG-TUM) [inline src/autopas/molecularDynamics/LJFunctor.h:858]:
```suggestion
    // checks whether particle I is in the domain box. Unused if calculateGlobals is false!
```
THe second part seems to be not true anymore.

File: src/autopas/molecularDynamics/LJFunctor.h [hunk 7]
```diff
@@ -1022,14 +939,16 @@ class LJFunctor
 
           const SoAFloatPrecision dr2 = drx2 + dry2 + drz2;
 
-          const SoAFloatPrecision mask = (dr2 <= cutoffsquare) ? 1. : 0.;
+          // ownedStateI is already checked.
+          const bool mask = dr2 <= cutoffsquare and
+                            /*ownedStateI != OwnershipState::dummy and */ ownedStateJ != OwnershipState::dummy;
```
Comment (FG-TUM) [inline src/autopas/molecularDynamics/LJFunctor.h:944]: same again

File: src/autopas/particles/ParticleBase.h [hunk 8]
```diff
@@ -155,18 +159,42 @@ class ParticleBase {
...
+
+  /**
+   * Marks a particle as deleted.
+   */
+  void markAsDeleted() {
+    // Set ownership as dummy.
+    _ownershipState = OwnershipState::dummy;
+    // Also mark position as very big, this prevents misuse in the force calculation.
+    //_r = {std::numeric_limits<floatType>::max(), std::numeric_limits<floatType>::max(),
+    //      std::numeric_limits<floatType>::max()};
```
Comment (FG-TUM) [inline src/autopas/particles/ParticleBase.h:191]: if it is not needed anymore remove it

File: src/autopas/sph/SPHCalcDensityFunctor.h [hunk 9]
```diff
@@ -138,8 +151,16 @@ class SPHCalcDensityFunctor : public Functor<Particle, ParticleCell, typename Pa
     double *const __restrict__ smthptr2 = soa2.template begin<Particle::AttributeNames::smth>();
     double *const __restrict__ massptr2 = soa2.template begin<Particle::AttributeNames::mass>();
 
+    const auto *const __restrict__ ownedStatePtr1 = soa1.template begin<Particle::AttributeNames::ownershipState>();
+    const auto *const __restrict__ ownedStatePtr2 = soa2.template begin<Particle::AttributeNames::ownershipState>();
+
     size_t numParticlesi = soa1.getNumParticles();
     for (unsigned int i = 0; i < numParticlesi; ++i) {
+      // checks whether particle 1 is in the domain box, unused if calculateGlobals is false!
```
Comment (FG-TUM) [inline src/autopas/sph/SPHCalcDensityFunctor.h:159]: same as above

File: src/autopas/sph/SPHCalcDensityFunctor.h [hunk 10]
```diff
@@ -180,6 +204,13 @@ class SPHCalcDensityFunctor : public Functor<Particle, ParticleCell, typename Pa
                         bool newton3) override {
     if (soa.getNumParticles() == 0) return;
 
+    const auto *const __restrict__ ownedStatePtr = soa.template begin<Particle::AttributeNames::ownershipState>();
+
+    // checks whether particle 1 is in the domain box, unused if calculateGlobals is false!
```
Comment (FG-TUM) [inline src/autopas/sph/SPHCalcDensityFunctor.h:209]: again

File: src/autopas/sph/SPHCalcHydroForceFunctor.h [hunk 11]
```diff
@@ -106,32 +109,38 @@ class SPHCalcHydroForceFunctor : public Functor<SPHParticle, FullParticleCell<SP
...
+    double *const __restrict__ velYptr = soa.template begin<Particle::AttributeNames::velY>();
+    double *const __restrict__ velZptr = soa.template begin<Particle::AttributeNames::velZ>();
+    double *const __restrict__ accXptr = soa.template begin<Particle::AttributeNames::accX>();
+    double *const __restrict__ accYptr = soa.template begin<Particle::AttributeNames::accY>();
+    double *const __restrict__ accZptr = soa.template begin<Particle::AttributeNames::accZ>();
+
+    const auto *const __restrict__ ownedStatePtr = soa.template begin<Particle::AttributeNames::ownershipState>();
 
     for (unsigned int indexFirst = 0; indexFirst < soa.getNumParticles(); ++indexFirst) {
+      // checks whether particle 1 is in the domain box, unused if calculateGlobals is false!
```
Comment (FG-TUM) [inline src/autopas/sph/SPHCalcHydroForceFunctor.h:139]: again

File: src/autopas/sph/SPHCalcHydroForceFunctor.h [hunk 12]
```diff
@@ -218,53 +227,56 @@ class SPHCalcHydroForceFunctor : public Functor<SPHParticle, FullParticleCell<SP
...
+    double *const __restrict__ velZptr2 = soa2.template begin<Particle::AttributeNames::velZ>();
+    double *const __restrict__ accXptr2 = soa2.template begin<Particle::AttributeNames::accX>();
+    double *const __restrict__ accYptr2 = soa2.template begin<Particle::AttributeNames::accY>();
+    double *const __restrict__ accZptr2 = soa2.template begin<Particle::AttributeNames::accZ>();
+
+    const auto *const __restrict__ ownedStatePtr1 = soa1.template begin<Particle::AttributeNames::ownershipState>();
+    const auto *const __restrict__ ownedStatePtr2 = soa2.template begin<Particle::AttributeNames::ownershipState>();
 
     for (unsigned int indexFirst = 0; indexFirst < soa1.getNumParticles(); ++indexFirst) {
+      // checks whether particle 1 is in the domain box, unused if calculateGlobals is false!
```
Comment (FG-TUM) [inline src/autopas/sph/SPHCalcHydroForceFunctor.h:275]: and again

File: src/autopas/sph/SPHCalcHydroForceFunctor.h [hunk 13]
```diff
@@ -362,32 +374,38 @@ class SPHCalcHydroForceFunctor : public Functor<SPHParticle, FullParticleCell<SP
...
-    double *const __restrict__ zptr = soa.template begin<autopas::sph::SPHParticle::AttributeNames::posZ>();
-    double *const __restrict__ velXptr = soa.template begin<autopas::sph::SPHParticle::AttributeNames::velX>();
-    double *const __restrict__ velYptr = soa.template begin<autopas::sph::SPHParticle::AttributeNames::velY>();
-    double *const __restrict__ velZptr = soa.template begin<autopas::sph::SPHParticle::AttributeNames::velZ>();
-    double *const __restrict__ accXptr = soa.template begin<autopas::sph::SPHParticle::AttributeNames::accX>();
-    double *const __restrict__ accYptr = soa.template begin<autopas::sph::SPHParticle::AttributeNames::accY>();
-    double *const __restrict__ accZptr = soa.template begin<autopas::sph::SPHParticle::AttributeNames::accZ>();
+    const auto *const __restrict__ ownedStatePtr = soa.template begin<Particle::AttributeNames::ownershipState>();
+
+    // checks whether particle 1 is in the domain box, unused if calculateGlobals is false!
```
Comment (FG-TUM) [inline src/autopas/sph/SPHCalcHydroForceFunctor.h:379]: lalala

File: tests/testAutopas/tests/molecularDynamics/LJFunctorAVXTest.cpp [hunk 14]
```diff
@@ -56,6 +56,12 @@ bool LJFunctorAVXTest::particleEqual(Particle &p1, Particle &p2) {
   EXPECT_EQ(p1.getID(), p2.getID());
 
   double tolerance = 1e-8;
+
+  /*
+std::cout << "comparing:" << std::endl
+            << "P1: " << p1.toString() << std::endl
+            << "P2: " << p2.toString() << std::endl;
+*/
```
Comment (FG-TUM) [inline tests/testAutopas/tests/molecularDynamics/LJFunctorAVXTest.cpp:64]: remove

File: tools/autopasTools/generators/RandomGenerator.h [hunk 15]
```diff
@@ -8,6 +8,7 @@
 
 #include <array>
 #include <functional>
+#include <iostream>
```
Comment (FG-TUM) [inline tools/autopasTools/generators/RandomGenerator.h:11]: is this needed?

**Pull request merged**